### PR TITLE
C interface for progress functions

### PR DESCRIPTION
**Pull request by ares201005**
C test for density matrix module (passed!)
Python script to generate c wrappers and headers
Add c interface for prg_chebyshev_mod module (Tests TBA) 
Add c interface for prg_charges_mod (Tests TBA)

PS: 
1) I found I cannot re-pen previous PR, so created a new PR (will not close it this time). 
2) There are some ci tests fail, I will continue working on them.  In the meanwhile, please comment on the code. 
3) I am working on the rest of the modules and corresponding tests as well, will push them into this PR once ready.

- - -
This change is [Reviewable](https://reviewable.io/reviews/lanl/qmd-progress/262)

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -88,6 +102,13 @@ if(PROGRESS_GRAPHLIB)
   progress_test(prg_build_zmatGP)
 endif()
 
+# c test modules
+progress_test(prg_density_c)
```
Comment (ares201005) [inline tests/CMakeLists.txt:106]: All the c test names end with "_c" in order to avoid conflicts with Fortran test names.

**Comment by ares201005** [comment]
Now all the CI checks passed!

File: benchmarks/dmconstruction/dmconstruction.F90 [hunk 2]
```diff
@@ -29,7 +28,7 @@ program hmodel
   real(dp) :: ef,sparsity,dec,mlsi,mlsf,bnorm
   character(20) :: bml_dmode
 
-  integer, external :: iargc
+  !integer, external :: iargc
```
Comment (jeanlucf22) [inline benchmarks/dmconstruction/dmconstruction.F90:31]: remove

File: examples/c_interface/progress.c [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
+#include "prg_progress_mod.h"
+
+int
+main(
+    )
+{
+    //prg_progress_init_c();   // initialize progress
```
Comment (jeanlucf22) [inline examples/c_interface/progress.c:8]: why keep these commented lines?

File: examples/c_interface/scf.c [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#include <stdlib.h>
```
Comment (jeanlucf22) [inline examples/c_interface/scf.c:1]: what is this code supposed to do? Is 'scf' a good name for it? Please add comments

File: scripts/transform.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
+import re
```
Comment (jeanlucf22) [inline scripts/transform.py:1]: Please add comments describing what this code is doing

File: scripts/transform.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+   print("Public subroutines are \n")
+   for subroutine in public_subroutines:
+       print(f"  public :: {subroutine}_c")
+   print("")
+   return public_subroutines
+
+
+def transform_fortran_file(filename):
+
+    #subroutine_re = re.compile(r'subroutine\s+([a-zA-Z0-9_]+)\s*\((.*)\)')
```
Comment (jeanlucf22) [inline scripts/transform.py:30]: remove commented lines unless they have a purpose/could be useful later

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -49,14 +49,26 @@ add_library(progress
   prg_timer_mod.F90
   prg_xlbo_mod.F90
   prg_graphsolver_mod.F90
-  prg_xlbokernel_mod.F90)
+  prg_xlbokernel_mod.F90
+  prg_c_interface.F90)
+  #prg_densitymatrix_mod_c.F90)
```
Comment (jeanlucf22) [inline src/CMakeLists.txt:54]: remove commented lines

File: src/prg_c_interface.F90 [hunk 8]
```diff
@@ -0,0 +1,1404 @@
...
+    ham_bml%ptr = ham_bml_c
+    rho_bml%ptr = rho_bml_c
+
+    call prg_build_density_T0(ham_bml, rho_bml, threshold, bndfil, eigenvalues)
+    eigenvalues_out = eigenvalues
+    deallocate(eigenvalues)
+
+  end subroutine prg_build_density_T0_c
+
+  subroutine prg_build_density_T_c(norbs, ham_bml_c, rho_bml_c, threshold, bndfil, kbt, ef, eigenvalues_out) bind(C, name="prg_build_density_T")
```
Comment (jeanlucf22) [inline src/prg_c_interface.F90:74]: could you split these long lines?

File: src/prg_c_interface.F90 [hunk 9]
```diff
@@ -0,0 +1,1404 @@
...
+  !------------------------------------------------
+
+  subroutine prg_get_nameandext_c(fullfilename, filename, ext) bind(C, name="prg_get_nameandext")
+    character(c_char), value :: fullfilename
+    character(c_char), value :: filename
+    character(c_char), value :: ext
+    call prg_get_nameandext(fullfilename, filename, ext)
+  end subroutine prg_get_nameandext_c
+
+  !  subroutine prg_parse_system_c(system, filename, extin) bind(C, name="prg_parse_system")
```
Comment (jeanlucf22) [inline src/prg_c_interface.F90:733]: are all these commented lines useful?

File: tests/src/main.c [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    char bml_type[21];          // 20 char + null termination '\0'
+    char sp2conv_sue[21];       // 20 char + null termination '\0'
+    char dummy[10][21];         // 20 char + null termination '\0' each
+
+    double threshold, gthreshold, idempotency;
+    double sp2tol, idempotency_tol;
+    double bndfil, tscale, tracelimit, beta;
+    double error_calc, error_tol, errlimit;
+
+    /*
```
Comment (jeanlucf22) [inline tests/src/main.c:46]: remove commented lines

File: tests/src/main.c [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    precision = double_real;
+    norb = 600;
+    threshold = 1.0e-9;
+    bndfil = 0.666666666666666666;
+    idempotency_tol = 1.0e-8;
+
+    double kbt = 0.01;
+    double mu = 0.0;
+    double scale_factor = 0.5;
+    double *eigenvalues = malloc(norb * sizeof(double));
```
Comment (jeanlucf22) [inline tests/src/main.c:89]: I don't see a free() corresponding to this malloc

**Review by jeanlucf22 — CHANGES_REQUESTED** [review]

File: examples/c_interface/scf.c (hunk 4 above)
Comment (jeanlucf22) [reply examples/c_interface/scf.c:1]: I still do not understand what this file is about. Maybe it should not be added yet?

**Comment by ares201005** [comment]
Changes: 1) Removed scf.c and other commented lines; 2) Added sp2 tests.

**Comment by ares201005** [comment]
hmm, two CI tests fail on sp2, will work on the fix.

**Review by jeanlucf22 — APPROVED** [review]

File: scripts/transform.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+headers in prg_progress_mod.h.
+
+Note: the generated wrappers are not bug-free, it is only used to facilitate building
+c interrface.
+"""
+
+def get_public(fortran_code):
+
+   # Define the regex pattern to match lines starting with 'public ::'
+   pattern = re.compile(r'public ::(.*)')
```
Comment (nicolasbock) [inline scripts/transform.py:19]:
```suggestion
   pattern = re.compile(r'public *::(.*)')
```
The standard does not require exactly one whitespace after the public keyword.

**Comment by ares201005** [comment]
Now this PR can be merged. Thanks.

**Review by nicolasbock — APPROVED** [review]

**Pull request merged**